src/realm/realm.h

```diff
@@ -0,0 +1,1612 @@
+    RLM_ERR_MODIFY_PRIMARY_KEY,
+    RLM_ERR_READ_ONLY_PROPERTY,
+    RLM_ERR_PROPERTY_NOT_NULLABLE,
+    RLM_ERR_INVALID_ARGUMENT,
+
+    RLM_ERR_LOGIC,
+    RLM_ERR_NO_SUCH_TABLE,
+    RLM_ERR_TABLE_NAME_IN_USE,
+    RLM_ERR_CROSS_TABLE_LINK_TARGET,
+    RLM_ERR_DESCRIPTOR_MISMATCH,
+    RLM_ERR_UNSUPPORTED_FILE_FORMAT_VERSION,
+    RLM_ERR_MULTIPLE_SYNC_AGENTS,
+    RLM_ERR_ADDRESS_SPACE_EXHAUSTED,
+    RLM_ERR_MAXIMUM_FILE_SIZE_EXCEEDED,
+    RLM_ERR_OUT_OF_DISK_SPACE,
+    RLM_ERR_KEY_NOT_FOUND,
+    RLM_ERR_COLUMN_NOT_FOUND,
+    RLM_ERR_COLUMN_ALREADY_EXISTS,
+    RLM_ERR_KEY_ALREADY_USED,
+    RLM_ERR_SERIALIZATION_ERROR,
+    RLM_ERR_INVALID_PATH_ERROR,
+    RLM_ERR_DUPLICATE_PRIMARY_KEY_VALUE,
+
+    RLM_ERR_INDEX_OUT_OF_BOUNDS,
+
+    RLM_ERR_INVALID_QUERY_STRING,
+    RLM_ERR_INVALID_QUERY,
+    // ...
+} realm_errno_e;
+
+typedef enum realm_logic_error_kind {
+    RLM_LOGIC_ERR_NONE = 0,
+    RLM_LOGIC_ERR_STRING_TOO_BIG,
+    // ...
+} realm_logic_error_kind_e;
+
+typedef struct realm_error {
+    realm_errno_e error;
+    realm_string_t message;
+    union {
+        int code;
+        realm_logic_error_kind_e logic_error_kind;
+    } kind;
+} realm_error_t;
+
+/* Schema types */
+
+typedef enum realm_column_attr {
+    // Values matching `realm::ColumnAttr`.
+    RLM_COLUMN_ATTR_NONE = 0,
+    RLM_COLUMN_ATTR_INDEXED = 1,
+    RLM_COLUMN_ATTR_UNIQUE = 2,
+    RLM_COLUMN_ATTR_RESERVED = 4,
+    RLM_COLUMN_ATTR_STRONG_LINKS = 8,
+    RLM_COLUMN_ATTR_NULLABLE = 16,
+    RLM_COLUMN_ATTR_LIST = 32,
+    RLM_COLUMN_ATTR_DICTIONARY = 64,
+    RLM_COLUMN_ATTR_COLLECTION = 64 + 32,
+} realm_column_attr_e;
+
+typedef enum realm_property_type {
+    // Values matching `realm::ColumnType`.
+    RLM_PROPERTY_TYPE_INT = 0,
+    RLM_PROPERTY_TYPE_BOOL = 1,
+    RLM_PROPERTY_TYPE_STRING = 2,
+    RLM_PROPERTY_TYPE_BINARY = 4,
+    RLM_PROPERTY_TYPE_MIXED = 6,
+    RLM_PROPERTY_TYPE_TIMESTAMP = 8,
+    RLM_PROPERTY_TYPE_FLOAT = 9,
+    RLM_PROPERTY_TYPE_DOUBLE = 10,
+    RLM_PROPERTY_TYPE_DECIMAL128 = 11,
+    RLM_PROPERTY_TYPE_OBJECT = 12,
+    RLM_PROPERTY_TYPE_LINKING_OBJECTS = 14,
+    RLM_PROPERTY_TYPE_OBJECT_ID = 15,
+    RLM_PROPERTY_TYPE_UUID = 17,
+} realm_property_type_e;
+
+typedef enum realm_collection_type {
+    RLM_COLLECTION_TYPE_NONE = 0,
+    RLM_COLLECTION_TYPE_LIST = 1,
+    RLM_COLLECTION_TYPE_SET = 2,
+    RLM_COLLECTION_TYPE_DICTIONARY = 4,
+} realm_collection_type_e;
+
+typedef struct realm_property_info {
+    realm_string_t name;
+    realm_string_t public_name;
+    realm_property_type_e type;
+    realm_collection_type_e collection_type;
+
+    realm_string_t link_target;
+    realm_string_t link_origin_property_name;
+    realm_col_key_t key;
+    int flags;
+} realm_property_info_t;
+
+typedef struct realm_class_info {
+    realm_string_t name;
+    realm_string_t primary_key;
+    size_t num_properties;
+    size_t num_computed_properties;
+    realm_table_key_t key;
+    int flags;
+} realm_class_info_t;
```

Comment:
Because it is used to communicate the number of properties, so the caller can allocate space for them (this choice comes from wanting to avoid a `realm_destroy_class_info()` function to clean up some internal list).

Reply:
I don't understand this argument - please elaborate.

Reply:
The caller is responsible for allocating space for the array of `realm_property_info_t`, so the caller needs to know how many there are.

Getting a full view of the schema for a class then involves code such as:

```c
realm_class_info_t clsinfo;
size_t num_properties;
realm_find_class(rlm, "Foo", NULL, &clsinfo);

realm_property_info_t* properties = malloc(sizeof(realm_property_info_t) * clsinfo.num_properties);
realm_get_class_properties(rlm, clsinfo.key, properties, clsinfo.num_properties, NULL);

// ...

free(properties);
```

The alternative is to turn `realm_class_info_t` into an object-like structure, but that's not great because it implies allocating space for the properties, even if the caller is only interested in stuff like the class's table key, primary key name, or flags.

Reply:
I can see you point in being more explicit in the ownership of the memory. Another approach could have been to have an interface where you could get information on one property at a time. That would support not being required to allocate an array from the heap. Like it is done for the class info.

Reply:
I would be fine with adding an index-based API for this if there is popular demand. :-)